File: leetcode-cn/0743.0_Network_Delay_Time.py
'''
Dijkstra(Greedy)

执行用时：68 ms, 在所有 Python3 提交中击败了93.74% 的用户
内存消耗：16.1 MB, 在所有 Python3 提交中击败了79.04% 的用户
'''
class Solution:
    def networkDelayTime(self, times: List[List[int]], n: int, k: int) -> int:
        # graph
        g = [[float('inf')] * n for _ in range(n)]
        # used 
        used = [False] * n
        # dist
        dist = [float('inf')] * n
        dist[k - 1] = 0  # source
        # set graph
        for u, v, w in times:
            g[u - 1][v - 1] = w
        
        for _ in range(n):
            # to find the nearest unused node x
            x = -1
            for y, u in enumerate(used):
                if not u and (x == -1 or dist[y] < dist[x]):
                    x = y
            used[x] = True
            for v, w in enumerate(g[x]):
                dist[v] = min(dist[v], dist[x] + w)

        ans = max(dist)
        return ans if ans < float('inf') else -1        

'''
Dijkstra + visited

执行用时：64 ms, 在所有 Python3 提交中击败了87.77% 的用户
内存消耗：16.1 MB, 在所有 Python3 提交中击败了93.85% 的用户
通过测试用例：52 / 52
'''
class Solution:
    def networkDelayTime(self, times: List[List[int]], n: int, k: int) -> int:
        # graph
        g = defaultdict(dict)
        for u, v, w in times:
            g[u - 1][v - 1] = w
        used = [False] * n 
        dist = [float('inf')] * n 
        dist[k - 1] = 0

        for _ in range(n):
            # to find nearest unused node
            x = -1
            for i, use in enumerate(used):
                if not use and (x == -1 or dist[i] < dist[x]):
                    x = i
            used[x] = True
            for child, w in g[x].items():
                dist[child] = min(dist[child], dist[x] + w)
    
        m = max(dist)
        return [m, -1][m == float('inf')]

'''
Dijkstra + heapq

执行用时：68 ms, 在所有 Python3 提交中击败了77.51% 的用户
内存消耗：16.3 MB, 在所有 Python3 提交中击败了24.97% 的用户
通过测试用例：52 / 52
'''
class Solution:
    def networkDelayTime(self, times: List[List[int]], n: int, k: int) -> int:
        # graph
        g = defaultdict(dict)
        for u, v, w in times:
            g[u - 1][v - 1] = w
        dist = [float('inf')] * n 
        dist[k - 1] = 0
        # heap (cost, node)
        h = [(0, k - 1)]
        while h:
            cost, node = heappop(h)
            for child, time in g[node].items():
                if cost + time < dist[child]:
                    dist[child] = cost + time
                    heappush(h, (dist[child], child))
                
        m = max(dist)
        return [m, -1][m == float('inf')]

File: cpp/0743.0_Network_Delay_Time.cpp
/*
不使用优先队列，适用于稠密图：

    不使用优先队列，但引入visited向量记录是否使用过了（使用过的意思是被当做中间节点 ）
    使用邻接矩阵表示邻接关系

执行用时：108 ms, 在所有 C++ 提交中击败了46.33% 的用户
内存消耗：40.3 MB, 在所有 C++ 提交中击败了18.80% 的用户
通过测试用例：52 / 52
*/

class Solution {
public:
    int networkDelayTime(vector<vector<int>>& times, int n, int k) {
        vector<vector<long long>> graph(n, vector<long long>(n, INT_MAX));
        for (int i = 0; i < n; ++i) {
            graph[i][i] = 0;
        }
        for (auto t: times) {
            graph[t[0] - 1][t[1] - 1] = t[2];
        }
        vector<bool> visited(n, false);

        for (int i = 0; i < n; i++) {
            // to find nearest unused node
            int min_id = -1;
            for (int j = 0; j < n; j++) {
                if (visited[j] == false && (min_id == -1 || graph[k - 1][j] < graph[k - 1][min_id])) {
                    min_id = j;
                }
            }
            visited[min_id] = true;
            // update dist
            for (int j = 0; j < n; ++j) {
                if (graph[k - 1][min_id] + graph[min_id][j] < graph[k - 1][j]) {
                    graph[k - 1][j] = graph[k - 1][min_id] + graph[min_id][j];
                }
            }
        }

        int ans = 0;
        for (int i = 0; i < n; ++i) {
            if (graph[k - 1][i] == INT_MAX) {
                return -1;
            }
            ans = max(ans, (int)graph[k - 1][i]);
        }
        return ans;
    }
};
